pkg/controllers/namespace/namespace_controller.go

```diff
@@ -0,0 +1,225 @@
+    If the ns contains this annotation, first check if	the namespace in annotation is the same as
+    the one in ns event, if yes, create an infra VPC for it. if	not, skip the whole ns event as the infra
+    VPC will be created its corresponding ns creation event.
+  - "nsx.vmware.com/vpc_network_config":"<Supervisor ID>"
+    If ns do not contains "nsx.vmware.com/vpc_name" annotation. Use this annotation to handle VPC creation.
+    VPC will locate the network config with the CR name, and create VPC using its config.
+  - If the ns do not have either of the annotation above, then we believe it is using default VPC, try to search
+    default VPC in network config CR store. The default VPC network config CR's name is "default".
+*/
+func (r *NamespaceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+
+	obj := &v1.Namespace{}
+	log.Info("reconciling K8s namespace resources", "namespace", req.NamespacedName)
+	metrics.CounterInc(r.NSXConfig, metrics.ControllerSyncTotal, common.MetricResTypeNamespace)
+
+	if err := r.Client.Get(ctx, req.NamespacedName, obj); err != nil {
+		log.Error(err, "unable to fetch namespace.", "req", req.NamespacedName)
+		return common.ResultRequeueAfter5mins, err
+	}
+
+	// processing create/update event
+	ns := obj.GetName()
+	if obj.ObjectMeta.DeletionTimestamp.IsZero() {
+		metrics.CounterInc(r.NSXConfig, metrics.ControllerUpdateTotal, common.MetricResTypeNamespace)
+		log.Info("start processing namespace create event", "namespace", ns)
+		ctx := context.Background()
+		annotations := obj.GetAnnotations()
+		// read anno "nsx.vmware.com/vpc_name", if ns contains this annotation, it means it will share
+		// infra VPC, if the ns in the annotation is the same as ns event, create infra VPC, if not,
+		// skip the event.
+		nc_name, nc_exist := annotations[NS_ANNO_VPC_NC_NAME]
+		vpc_name, name_exist := annotations[NS_ANNO_VPC_NAME]
+		if name_exist {
+			res := strings.Split(vpc_name, "/")
+			// The format should be namespace/vpc_name
+			if len(res) != 2 {
+				message := fmt.Sprintf("incorrect vpc_name annotation %s for namespace %s.", vpc_name, ns)
+				r.namespaceError(&ctx, obj, message, nil)
+				// If illegal format, skip handling this event?
+				return common.ResultNormal, nil
+			}
+			if ns == res[0] {
+				// namespace name is the same as ns creation event, create VPC CR
+				// read infra network config cr name which is in annotation nsx.vmware.com/vpc-network-configuration
+				if !nc_exist {
+					// if no infra network config cr annotation, skip the event
+					message := fmt.Sprintf("missing network config CR name annotation for namespace %s.", ns)
+					r.namespaceError(&ctx, obj, message, nil)
+					return common.ResultNormal, nil
+				}
+				//TODO: use indicated vpc name to create infra VPC: res[1]
+				if _, err := r.createVPCCR(&ctx, obj, ns, nc_name, &vpc_name); err != nil {
+					message := "failed to create VPC CR for infra, requeue ns creation event"
```

Comment:
I don't think we need to handle infra case specifically, I mean you can decide whether to create VPC CR here, then create it in the end together in line 201